Allow memecoins for futures in the MarketCapPairList plugin

mrpabloyeah opened the pull request:
<!-- Thank you for sending your pull request. But first, have you included
unit tests, and is your code PEP8 conformant? [More details](https://github.com/freqtrade/freqtrade/blob/develop/CONTRIBUTING.md)
-->
## Summary

Currently, MarketCapPairList ignores pairs containing the prefix "1000" (in Binance Futures) or "k" (in Hyperliquid Futures) because it cannot find the coin name in CoinGecko and therefore cannot obtain the market cap of these coins. 

This PR addresses this issue by removing the prefix when searching for the coin name and replacing it when creating the pair list. This way, the MarketCapPairList plugin now supports memecoins like PEPE, SHIB, and BONK for futures trading.

mrpabloyeah committed:
Allow memecoins in MarketCapPairList plugin

xmatthias commented on freqtrade/plugins/pairlist/MarketCapPairList.py:
```diff
@@ -147,6 +147,21 @@ def gen_pairlist(self, tickers: Tickers) -> list[str]:
+        :return: Normalized pair string.
+        """
+
+        if pair.startswith("1000"):  # Binance
```
I don't think that's correct / desired.

1000PEPE is a different market than PEPE ... both are valid markets though - so trimming one in favor of the other is not necessarily what you'd expect i think.

mrpabloyeah replied:
Well, to be precise, PEPE is not a market, it is the asset that you can trade in different markets or pairs like PEPE/USDT or 1000PEPE/USDT. This plugin sorts the pairlist based on the market capitalization of the asset (PEPE in this case), it is not based on the liquidity of the pair (PEPE/USDT or 1000PEPE/USDT), that's why I think if we are going to trade in the 1000PEPE/USDT pair we want the plugin to be based on the market capitalization of PEPE, since it is the asset that we are going to trade in this pair.

xmatthias replied:
> I've updated the title as this issue only occurs in futures trading.

Well i'm not sure that's true - at least the code doesn't suggest it's limited to futures - and from a pair/asset perspective, that's definitely not the case.


Running `freqtrade list-markets --exchange binance | grep 1000` reveals (for example) [1000SATS](https://www.binance.com/en/trade/1000SATS_USDC) (which is a valid pair also listed on [coingecko](https://www.coingecko.com/en/coins/1000sats-ordinals).

It's the same with 1000CAT - and probably a few others if we start to look at other exchanges (other than binance).
Sure, there's also futures markets without corresponding "non-1000" market - but this suggests that while it may add 1000PEPE - it'll also become problematic for 1000SATS, which will now use the wrong coingecko reference (and hence, a different sorting and category).

mrpabloyeah replied:
1000SATS exists on Coingecko, but you can't see the market cap for 1000SATS because it's not a coin; it simply represents 1000 units of the SATS coin. The same applies to 1000CAT or any coin with that prefix.

Currently the plugin ignores those pairs, but what it should do is get the market cap of the coin that is traded on that pair, and that is what this PR fixes.

xmatthias replied:
considering both 1000SATS/USDT and SATS/USDT exist - what'd be the outcome if i use this pairlist?

you'd have both pairs - as they have exactly the same market cap (it's the same value that's used under the hood).
i'm not entirely sure that corresponds to what i'd expect to happen in such a case.

personally, i think ignoring those pairs is hence the better behavior than having the same coin twice one after the other.

mrpabloyeah replied:
I don't think this is currently the case (although I'm not 100% sure), but I think that while the 1000SATS/USDT pair exists, the SATS/USDT pair doesn't exist on the same exchange for the same market (futures or spot). 

For example, on Binance Spot there is SHIB/USDT but there is no 1000SHIB/USDT, and on Binance Futures there is 1000SHIB/USDT but there is no SHIB/USDT.

xmatthias replied:
i think such a mapping is pretty risky - as it's based on assumptions - no?

assumptions based on (a couple) exchanges we can test  / verify - and could in reality change with any pair listing (i doubt exchanges have silently agreed and communicated to never list both 1000PEPE and PEPE on the same market ... ? ).

mrpabloyeah replied:
> (i doubt exchanges have silently agreed and communicated to never list both 1000PEPE and PEPE on the same market ... ? ).

Hahaha that was a good one

xmatthias commented on freqtrade/plugins/pairlist/MarketCapPairList.py:
```diff
@@ -147,6 +147,21 @@ def gen_pairlist(self, tickers: Tickers) -> list[str]:
+
+        if pair.startswith("1000"):  # Binance
+            return pair[4:]
+        elif pair.startswith("k"):  # Hyperliquid
```
any source this is true "globally" ? 
I understand it MAY be a limitation with hyperliquid - but can you generalize this that "if a coin starts with k, it's the same as the coin witout k"? 
(keep in mind - this applies to all coins across all exchanges - so it must be applicable to all)

mrpabloyeah replied:
As far as I know, the pair name is written in capital letters on all exchanges, so I thought that being a lowercase "k" shouldn't cause any conflict on other exchanges. However, if you think this might cause any conflict in the future, we can specify in the condition that the exchange must be hyperliquid.

xmatthias replied:
this one i'm not sure - it's "generally" a convention to have names all uppercase (source: https://docs.ccxt.com/#/README?id=market-structure)
> `    'symbol':  'BTC/USD',     // uppercase string literal of a pair of currencies`

I think this is hence a problem / bug in ccxt - not something we should "map around" ...

xmatthias replied:
https://github.com/ccxt/ccxt/pull/26332 will change this - and update it to KBONK (default is to uppercase the base name).

(with this in mind) -  we can't safely assume that every coin starting with "K" might correspond to the one without k - as KOBE/USDC is a valid pair - though OBE/USDC is not (maybe it's a valid pair - though it's definitely not the same pair).

mrpabloyeah commented:
I've updated the title as this issue only occurs in futures trading.

I've also updated the PR summary to explain in more detail what issue it addresses.

mrpabloyeah commented:
Well, we'll have to find a solution for this, right? Or are we going to leave it like this, ignoring pairs that have this prefix? Do you have any ideas?

xmatthias commented:
> Well, we'll have to find a solution for this, right? Or are we going to leave it like this, ignoring pairs that have this prefix? Do you have any ideas?

To be fair - i didn't see this as a problem so far - so i'm not convinced it needs fixing (which i think has become clear above :laughing: )

There's no guarantee that coingecko has all pairs - or that what coingecko considers as "CoinA" is also "CoinA" on the exchange (there's been a few instances where the same name was given twice in the past). The pairlist handles this (somewhat) - though based on the assumption that "the exchange will have listed the higher one" - which may or may not be true.
An exchange may list a coin as "coinA" - while on coingecko it's "CoinA1", too (there's no guarantee once you move past the top 20-50 coins, really).

The docs for this also somewhat [warn from this](https://www.freqtrade.io/en/latest/plugins/#marketcappairlist) (red warning box at the bottom)

> Coingecko often has duplicate symbols, where the same symbol is used for different coins. Freqtrade will use the symbol as is and try to search for it on the exchange. If the symbol exists - it will be used. Freqtrade will however not check if the intended symbol is the one coingecko meant. This can sometimes lead to unexpected results, especially on low volume coins or with meme coin categories.

---

for hyperliquid - i don't currently see a solution. 
You can't reliably determine that KBONK should be BONK - while ignoring that KOBE != OBE.
We could hard-map these - but it'll be wrong / need maintenance with every new pair listing (which i'll guarantee, will be outdated after the first pair listing).

for the 1000xxx coins - we could in theory have this mapping - with the condition that we'll rip it out should we get reports of an actual "1000x" coin which is now no longer detected (no point in having a workaround for 10 coins, if it brakes other coins).

mrpabloyeah commented:
> > Well, we'll have to find a solution for this, right? Or are we going to leave it like this, ignoring pairs that have this prefix? Do you have any ideas?
> 
> To be fair - i didn't see this as a problem so far - so i'm not convinced it needs fixing (which i think has become clear above 😆 )
> 
> There's no guarantee that coingecko has all pairs - or that what coingecko considers as "CoinA" is also "CoinA" on the exchange (there's been a few instances where the same name was given twice in the past). The pairlist handles this (somewhat) - though based on the assumption that "the exchange will have listed the higher one" - which may or may not be true. An exchange may list a coin as "coinA" - while on coingecko it's "CoinA1", too (there's no guarantee once you move past the top 20-50 coins, really).
> 
> The docs for this also somewhat [warn from this](https://www.freqtrade.io/en/latest/plugins/#marketcappairlist) (red warning box at the bottom)
> 
> > Coingecko often has duplicate symbols, where the same symbol is used for different coins. Freqtrade will use the symbol as is and try to search for it on the exchange. If the symbol exists - it will be used. Freqtrade will however not check if the intended symbol is the one coingecko meant. This can sometimes lead to unexpected results, especially on low volume coins or with meme coin categories.
> 
> for hyperliquid - i don't currently see a solution. You can't reliably determine that KBONK should be BONK - while ignoring that KOBE != OBE. We could hard-map these - but it'll be wrong / need maintenance with every new pair listing (which i'll guarantee, will be outdated after the first pair listing).
> 
> for the 1000xxx coins - we could in theory have this mapping - with the condition that we'll rip it out should we get reports of an actual "1000x" coin which is now no longer detected (no point in having a workaround for 10 coins, if it brakes other coins).

Well, I don't know if I should classify this as a problem, but at least for me it makes the plugin useless, since it can't provide a pairlist with the top100 marcketcap coins, which is what I wanted to use it for. 

But I honestly don't have any more arguments to continue defending this PR's approach, so I'm closing it. I'll open another one in the future if I come up with another way to address this.

xmatthias commented:
> for the 1000xxx coins - we could in theory have this mapping - with the condition that we'll rip it out should we get reports of an actual "1000x" coin which is now no longer detected (no point in having a workaround for 10 coins, if it brakes other coins).

well maybe it didn't come across this way - i'm "kinda" fine with the "1000" approach - though i don't see a way for hyperliquid at the moment - because their naming doesn't suggest if it's a "1000" pair - or not

mrpabloyeah commented:
> > for the 1000xxx coins - we could in theory have this mapping - with the condition that we'll rip it out should we get reports of an actual "1000x" coin which is now no longer detected (no point in having a workaround for 10 coins, if it brakes other coins).
> 
> well maybe it didn't come across this way - i'm "kinda" fine with the "1000" approach - though i don't see a way for hyperliquid at the moment - because their naming doesn't suggest if it's a "1000" pair - or not

I'd like to find a solution that's also viable for Hyperliquid, since that's the exchange I use for live trading. I can't use Binance because my strategies are for futures and I live in Europe, so I only use it occasionally for dry runs.

xmatthias commented:
> I'd like to find a solution that's also viable for Hyperliquid, since that's the exchange I use for live trading

Well as said - the real problem here is that the coin itself doesn't give hints about this behavior anymore (which is correct, as ccxt assumes all symbols are uppercased).
ideally, the exchange (!) would return a field "is 1000 BONK" - but that's not there as far as i can see.
